### ISSUE TITLE
Full Test Coverage for Local/Remote Torch Tensor/Variable Operations

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#1385]
## Background

PySyft is designed to override all of the operations of PyTorch to allow for its use in new environments, such as [remote execution](https://github.com/OpenMined/PySyft/blob/master/examples/Remote%20PyTorch%20using%20Virtual%20Worker.ipynb), [fixed precision execution](https://github.com/OpenMined/PySyft/pull/1378), or [execution over tensors which have multiple owners](https://mortendahl.github.io/2017/04/17/private-deep-learning-with-mpc/). To implement this efficiently, we've used automatic code generation to generate the code for such operations automatically. However, we have recently been finding a [large number of edge cases](https://github.com/OpenMined/PySyft/issues/1382) which our code generation does not yet support.

Thus, we would like to have 100% testing coverage over [all tensor operations that PyTorch provides](https://pytorch.org/docs/stable/tensors.html). Each test should demonstrate the correct computation of a PyTorch tensor operation using VirtualWorker objects. Each test should show the operation working correctly for both local and remote tensors and should be installed in the [torch_test.py](https://github.com/OpenMined/PySyft/blob/master/test/torch_test.py) file in which there are already several examples you can use as a reference.

So, to complete a Pull Request on this Issue:
1. If you have not yet done so, create a Fork of PySyft (by clicking the "Fork" button on the top right corner of this page)
2. Find a [PyTorch tensor operation](https://pytorch.org/docs/stable/tensors.html) that has not yet been tested in [torch_test.py](https://github.com/OpenMined/PySyft/blob/master/test/torch_test.py)
3. Write a test showing that the operation works on local tensors.
4. Write a test showing that the operation works on remote tensors (by calling .send() on each tensor... performing the op... then calling .get() and checking that the results are correct)
5. Repeat steps 2 and 3 for Variables.
6. Create a pull request and link to this Issue!

**Comment by alhparsa** [comment]
Here are all the methods for tensors:

remainder_, __getstate__, __rmul__, atan_, reinforce, __invert__, numpy, squeeze_, stride, type_as, index_fill, is_contiguous, frac, _set_index, baddbmm_, __eq__, cuda, norm, floor, masked_fill, __setattr__, __iadd__, index_add_, abs, normal_, sum, __idiv__, exp, is_leaf, grad_fn, resize_as_, t_, addcmul, atan2, scatter_add, ge, short, sinh_, long, sqrt_, trtrs, grad, volatile, __ne__, log1p, rsqrt_, symeig, _execution_engine, __dict__, tolist, addmv_, is_signed, masked_scatter, svd, register_hook, __sub__, addmv, __iand__, _write_metadata, __mul__, addmm_, median, new, add, __len__, dist, mul, mm, cos, __class__, sign, index, orgqr, div_, __rshift__, triu, potri, mv, renorm_, div, topk, requires_grad, __doc__, __new__, __dir__, __mod__, is_sparse, __or__, baddbmm, zeros_like, masked_copy_, __xor__, map_, sin, addbmm_, __gt__, lgamma, reciprocal, __imul__, eq, sub_, fmod_, remainder, __init_subclass__, _version, numel, split, expand, floor_, type, clamp_, bernoulli_, sigmoid_, _cdata, size, __ior__, addcdiv_, cosh, __array_wrap__, histc, ger, addcdiv, cumprod, ormqr, tril, kthvalue, __str__, sqrt, transpose_, permute, cat, cauchy_, tril_, __pow__, trace, __weakref__, trunc, __long__, apply_, is_same_size, _fallthrough_methods, __irshift__, gesv, pstrf, is_cuda, chunk, le_, view, ne_, __delattr__, masked_scatter_, asin, clone, min, _grad, byte, ndimension, output_nr, masked_copy, index_select, unsqueeze_, __getattribute__, __bool__, inverse, prod, _torch, char, __getitem__, cpu, data_ptr, __isub__, __and__, __int__, index_add, geqrf, __neg__, acos_, _unnarrow, __ixor__, zero_, sin_, set_, dim, unfold, __iter__, neg, renorm, __float__, __div__, __matmul__, bmm, scatter, log1p_, masked_fill_, sign_, qr, __getattr__, nonzero, reciprocal_, __delitem__, __lt__, rsqrt, bernoulli, atan, log, __ipow__, mean, __sizeof__, index_copy, __setstate__, pow_, uniform_, resize, ge_, __lshift__, transpose, __format__, dot, is_pinned, __subclasshook__, add_, exponential_, pin_memory, matmul, storage, sort, mul_, sigmoid, gt_, round, resize_, __rsub__, fill_, std, take, __le__, addcmul_, data, __setitem__, ne, erf, scatter_, narrow, copy_, var, get_device, is_set_to, addmm, pow, _advanced_index_select, btrisolve, nelement, __rdiv__, lgamma_, atan2_, index_copy_, trunc_, __truediv__, tan, __rtruediv__, map2_, __repr__, equal, lerp_, _get_type, mode, __init__, _check_advanced_indexing, share_memory_, lerp, expand_as, __itruediv__, erf_, sinh, erfinv_, storage_offset, clamp, cumsum, multinomial, ceil, gather, view_as, ones_like, eig, storage_type, resize_as, btrifact, gt, __add__, __reduce_ex__, shape, abs_, potrs, addbmm, retain_grad, geometric_, random_, half, acos, cos_, __ge__, detach_, _grad_fn, element_size, addr_, detach, index_fill_, int, __reduce__, __array__, __hash__, all, __nonzero__, select, contiguous, t, erfinv, repeat, tan_, addr, is_shared, max, tanh_, exp_, double, eq_, __radd__, lt, masked_select, __ilshift__, __rpow__, log_normal_, _advanced_index_add, neg_, scatter_add_, lt_, float, frac_, potrf, log_, gels, unsqueeze, cosh_, squeeze, asin_, put_, backward, ceil_, cross, diag, round_, triu_, le, _sparse_mask, tanh, fmod, __module__, __deepcopy__, _new_with_metadata_file, any, _backward_hooks, sub

**Comment by alhparsa** [comment]
Here is all the methods which can be applied to tensors both locally and remotely, if you have added a unit test for any of these put a check mark beside it so others will know which ones they can work one. Thanks in advance

- [ ] round_
- [ ] fmod
- [ ] __module__
- [ ] __deepcopy__
- [ ] _new_with_metadata_file
- [ ] any
- [ ] _backward_hooks
- [ ]sub
- [ ]remainder_
- [ ] __getstate__
- [ ] __rmul__
- [ ] atan_
- [ ] reinforce
- [ ] __invert__
- [ ] numpy
- [ ] squeeze_
- [ ] stride
- [ ] type_as
- [ ] index_fill
- [ ] is_contiguous
- [ ] frac
- [ ] _set_index
- [ ] baddbmm_
- [ ] __eq__
- [ ] cuda
- [ ] norm
- [ ] floor
- [ ] masked_fill
- [ ] __setattr__
- [ ] __iadd__
- [ ] index_add_
- [ ] abs
- [ ] normal_
- [ ] sum
- [ ] __idiv__
- [ ] exp
- [ ] is_leaf
- [ ] grad_fn
- [ ] resize_as_
- [ ] t_
- [ ] addcmul
- [ ] atan2
- [ ] scatter_add
- [ ] ge
- [ ] short
- [ ] sinh_
- [ ] long
- [ ] sqrt_
- [ ] trtrs
- [ ] grad
- [ ] volatile
- [ ] __ne__
- [ ] log1p
- [ ] rsqrt_
- [ ] symeig
- [ ] _execution_engine
- [ ] __dict__
- [ ] tolist
- [ ] addmv_
- [ ] is_signed
- [ ] masked_scatter
- [ ] svd
- [ ] register_hook
- [ ] __sub__
- [ ] addmv
- [ ] __iand__
- [ ] _write_metadata
- [ ] __mul__
- [ ] addmm_
- [ ] median
- [ ] new
- [ ] add
- [ ] __len__
- [ ] dist
- [ ] mul
- [ ] mm
- [ ] cos
- [ ] __class__
- [ ] sign
- [ ] index
- [ ] orgqr
- [ ] div_
- [ ] __rshift__
- [ ] triu
- [ ] potri
- [ ] mv
- [ ] renorm_
- [ ] div
- [ ] topk
- [ ] requires_grad
- [ ] __doc__
- [ ] __new__
- [ ] __dir__
- [ ] __mod__
- [ ] is_sparse
- [ ] __or__
- [ ] baddbmm
- [ ] zeros_like
- [ ] masked_copy_
- [ ] __xor__
- [ ] map_
- [ ] sin
- [ ] addbmm_
- [ ] __gt__
- [ ] lgamma
- [ ] reciprocal
- [ ] __imul__
- [ ] eq
- [ ] sub_
- [ ] fmod_
- [ ] remainder
- [ ] __init_subclass__
- [ ] _version
- [ ] numel
- [ ] split
- [ ] expand
- [ ] floor_
- [ ] type
- [ ] clamp_
- [ ] bernoulli_
- [ ] sigmoid_
- [ ] _cdata
- [ ] size
- [ ] __ior__
- [ ] addcdiv_
- [ ] cosh
- [ ] __array_wrap__
- [ ] histc
- [ ] ger
- [ ] addcdiv
- [ ] cumprod
- [ ] ormqr
- [ ] tril
- [ ] kthvalue
- [ ] __str__
- [ ] sqrt
- [ ] transpose_
- [ ] permute
- [ ] cat
- [ ] cauchy_
- [ ] tril_
- [ ] __pow__
- [ ] trace
- [ ] __weakref__
- [ ] trunc
- [ ] __long__
- [ ] apply_
- [ ] is_same_size
- [ ] _fallthrough_methods
- [ ] __irshift__
- [ ] gesv
- [ ] pstrf
- [ ] is_cuda
- [ ] chunk
- [ ] le_
- [ ] view
- [ ] ne_
- [ ] __delattr__
- [ ] masked_scatter_
- [ ] asin
- [ ] clone
- [ ] min
- [ ] _grad
- [ ] byte
- [ ] ndimension
- [ ] output_nr
- [ ] masked_copy
- [ ] index_select
- [ ] unsqueeze_
- [ ] __getattribute__
- [ ] __bool__
- [ ] inverse
- [ ] prod
- [ ] _torch
- [ ] char
- [ ] __getitem__
- [ ] cpu
- [ ] data_ptr
- [ ] __isub__
- [ ] __and__
- [ ] __int__
- [ ] index_add
- [ ] geqrf
- [ ] __neg__
- [ ] acos_
- [ ] _unnarrow
- [ ] __ixor__
- [ ] zero_
- [ ] sin_
- [ ] set_
- [ ] dim
- [ ] unfold
- [ ] __iter__
- [ ] neg
- [ ] renorm
- [ ] __float__
- [ ] __div__
- [ ] __matmul__
- [ ] bmm
- [ ] scatter
- [ ] log1p_
- [ ] masked_fill_
- [ ] sign_
- [ ] qr
- [ ] __getattr__
- [ ] nonzero
- [ ] reciprocal_
- [ ] __delitem__
- [ ] __lt__
- [ ] rsqrt
- [ ] bernoulli
- [ ] atan
- [ ] log
- [ ] __ipow__
- [ ] mean
- [ ] __sizeof__
- [ ] index_copy
- [ ] __setstate__
- [ ] pow_
- [ ] uniform_
- [ ] resize
- [ ] ge_
- [ ] __lshift__
- [ ] transpose
- [ ] __format__
- [ ] dot
- [ ] is_pinned
- [ ] __subclasshook__
- [ ] add_
- [ ] exponential_
- [ ] pin_memory
- [ ] matmul
- [ ] storage
- [ ] sort
- [ ] mul_
- [ ] sigmoid
- [ ] gt_
- [ ] round
- [ ] resize_
- [ ] __rsub__
- [ ] fill_
- [ ] std
- [ ] take
- [ ] __le__
- [ ] addcmul_
- [ ] data
- [ ] __setitem__
- [ ] ne
- [ ] erf
- [ ] scatter_
- [ ] narrow
- [ ] copy_
- [ ] var
- [ ] get_device
- [ ] is_set_to
- [ ] addmm
- [ ] pow
- [ ] _advanced_index_select
- [ ] btrisolve
- [ ] nelement
- [ ] __rdiv__
- [ ] lgamma_
- [ ] atan2_
- [ ] index_copy_
- [ ] trunc_
- [ ] __truediv__
- [ ] tan
- [ ] __rtruediv__
- [ ] map2_
- [ ] __repr__
- [ ] equal
- [ ] lerp_
- [ ] _get_type
- [ ] mode
- [ ] __init__
- [ ] _check_advanced_indexing
- [ ] share_memory_
- [ ] lerp
- [ ] expand_as
- [ ] __itruediv__
- [ ] erf_
- [ ] sinh
- [ ] erfinv_
- [ ] storage_offset
- [ ] clamp
- [ ] cumsum
- [ ] multinomial
- [ ] ceil
- [ ] gather
- [ ] view_as
- [ ] ones_like
- [ ] eig
- [ ] storage_type
- [ ] resize_as
- [ ] btrifact
- [ ] gt
- [ ] __add__
- [ ] __reduce_ex__
- [ ] shape
- [ ] abs_
- [ ] potrs
- [ ] addbmm
- [ ] retain_grad
- [ ] geometric_
- [ ] random_
- [ ] half
- [ ] acos
- [ ] cos_
- [ ] __ge__
- [ ] detach_
- [ ] _grad_fn
- [ ] element_size
- [ ] addr_
- [ ] detach
- [ ] index_fill_
- [ ] int
- [ ] __reduce__
- [ ] __array__
- [ ] __hash__
- [ ] all
- [ ] __nonzero__
- [ ] select
- [ ] contiguous
- [ ] t
- [ ] erfinv
- [ ] repeat
- [ ] tan_
- [ ] addr
- [ ] is_shared
- [ ] max
- [ ] tanh_
- [ ] exp_
- [ ] double
- [ ] eq_
- [ ] __radd__
- [ ] lt
- [ ] masked_select
- [ ] __ilshift__
- [ ] __rpow__
- [ ] log_normal_
- [ ] _advanced_index_add
- [ ] neg_
- [ ] scatter_add_
- [ ] lt_
- [ ] float
- [ ] frac_
- [ ] potrf
- [ ] log_
- [ ] gels
- [ ] unsqueeze
- [ ] cosh_
- [ ] squeeze
- [ ] asin_
- [ ] put_
- [ ] backward
- [ ] ceil_
- [ ] cross
- [ ] diag
- [ ] triu_
- [ ] le
- [ ] _sparse_mask
- [ ] tanh

**Comment by jvmancuso** [comment]
Heads up, some of those are deprecated and will raise exceptions (e.g. `reinforce`).  Also, some of these methods aren't meant to be called directly (e.g. the magic methods like `__radd__`, `__rpow__`, and so on).
